RFS/src/main/java/org/opensearch/migrations/bulkload/transformers/TransformerToIJsonTransformerAdapter.java

```diff
@@ -0,0 +1,165 @@
+    }
+
+    private void logTransformation(Map<String, Object> before, Map<String, Object> after) {
+        if (transformerLogger.isInfoEnabled()) {
+            try {
+                var transformationTuple = toTransformationMap(before, after);
+                var tupleString = MAPPER.writeValueAsString(transformationTuple);
+                transformerLogger.atInfo().setMessage("{}").addArgument(tupleString).log();
+            } catch (Exception e) {
+                log.atError().setCause(e).setMessage("Exception converting tuple to string").log();
+                transformerLogger.atInfo().setMessage("{ \"error\":\"{}\" }").addArgument(e::getMessage).log();
+                throw Lombok.sneakyThrow(e);
+            }
+        }
+    }
+
+    private Map<String, Object> toTransformationMap(Map<String, Object> before, Map<String, Object> after) {
+        var transformationMap = new LinkedHashMap<String, Object>();
+        transformationMap.put("before", before);
+        transformationMap.put("after", after);
+        return transformationMap;
+    }
+
+
+    @SuppressWarnings("unchecked")
+    private static Map<String, Object> objectNodeToMap(Object node) {
+        return (Map<String, Object>) MAPPER.convertValue(node, Map.class);
+    }
+
+    @SneakyThrows
+    private static String printMap(Map<String, Object> map) {
+        return MAPPER.writeValueAsString(map);
+    }
+
+    @SuppressWarnings("unchecked")
+    private MigrationItem transformMigrationItem(MigrationItem migrationItem) {
+        // Keep untouched original for logging
+        final Map<String, Object> originalMap = MAPPER.convertValue(migrationItem, Map.class);
+        var transformedMigrationItem = transformer.transformJson(MAPPER.convertValue(migrationItem, Map.class));
+        logTransformation(originalMap, transformedMigrationItem);
+        return MAPPER.convertValue(transformedMigrationItem, MigrationItem.class);
+    }
+
+    void updateTemplates(Collection<? extends MigrationItem> transformedItems, ObjectNode itemsRoot) {
+        itemsRoot.removeAll();
+        transformedItems.forEach(item ->
+                {
+                    log.atInfo().setMessage("Setting new item of type {}, name {}, body {}")
+                            .addArgument(item.type)
+                            .addArgument(item.name)
+                            .addArgument(item.body)
+                            .log();
+                    itemsRoot.set(item.name, item.body);
+                }
+        );
+    }
+
+    @Override
+    public GlobalMetadata transformGlobalMetadata(GlobalMetadata globalData) {
+        var inputJson = objectNodeToMap(globalData.toObjectNode());
+        log.atInfo().setMessage("BeforeJsonGlobal: {}").addArgument(() -> printMap(inputJson)).log();
+        var afterJson = transformer.transformJson(inputJson);
+        log.atInfo().setMessage("AfterJsonGlobal: {}").addArgument(() -> printMap(afterJson)).log();
```

Comment:
It's interesting to have a pre-processor transform for the global metadata but not the index settings.  Is this an opinionated stance or just a matter of current necessity?  Also - how does this fit into the bigger picture we previously discussed of having user-supplied bulk pre/post processing transforms then line-item transforms for each individual MigrationItem between those two steps?

Reply:
I've cleaned this up in the latest version, the unit of transformation for index settings is the entire object, whereas for global metadata it's the sub-object (individual template).